sched/clock/clock_perf.c

```diff
@@ -0,0 +1,176 @@
+#include <nuttx/arch.h>
+#include <nuttx/wdog.h>
+
+#if defined(CONFIG_PERF_OVERFLOW_CORRECTION) && ULONG_MAX != UINT64_MAX
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct perf_s
+{
+  struct wdog_s wdog;
+  unsigned long last;
+  unsigned long overflow;
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct perf_s g_perf;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * perf_update
+ ****************************************************************************/
+
+static void perf_update(wdparm_t arg)
+{
+  clock_t tick = (clock_t)LONG_MAX * TICK_PER_SEC / up_perf_getfreq();
+
+  perf_gettime();
+  wd_start((FAR struct wdog_s *)arg, tick, perf_update, arg);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * perf_init
+ ****************************************************************************/
+
+void perf_init(void)
+{
+  FAR struct perf_s *perf = &g_perf;
+  clock_t tick = (clock_t)LONG_MAX * TICK_PER_SEC / up_perf_getfreq();
+
+  perf->last = up_perf_gettime();
+
+  /* Periodic check for overflow */
+
+  wd_start(&perf->wdog, tick, perf_update, (wdparm_t)perf);
+}
+
+/****************************************************************************
+ * perf_gettime
+ ****************************************************************************/
+
+clock_t perf_gettime(void)
+{
+  FAR struct perf_s *perf = &g_perf;
+  unsigned long now = up_perf_gettime();
+
+  /* Check if overflow */
+
+  if (now < perf->last)
+    {
+      perf->overflow++;
+    }
+
+  perf->last = now;
+  return (clock_t)now | (clock_t)perf->overflow << 32;
+}
+
+/****************************************************************************
+ * perf_convert
+ ****************************************************************************/
+
+void perf_convert(clock_t elapsed, FAR struct timespec *ts)
+{
+  unsigned long freq = up_perf_getfreq();
+
+  ts->tv_sec  = elapsed / freq;
+  elapsed -= ts->tv_sec * freq;
+  ts->tv_nsec = NSEC_PER_SEC * elapsed / freq;
+}
+
+#elif defined(CONFIG_ALARM_ARCH) || defined (CONFIG_TIMER_ARCH) || \
+      defined(CONFIG_ARCH_PERF_EVENTS)
+
+/****************************************************************************
+ * perf_init
+ ****************************************************************************/
+
+void perf_init(void)
+{
+}
```

Comment:
shouldn't we call `up_perf_init()` here ? then arch specific logic doesn't have to worry about it anymore

Reply:
The argument is hard to decide, or let's change the argument of up_perf_init to void?

Reply:
We still need to provide information about CPU freq to up_perf_init. How we do that ? 
Can we use the common CPU frequency definition for boards ? something like `BOARD_SYSTICK_CLOCK` or `BOARD_SYSCLK_FREQUENCY`.

Reply:
or better - add an interface for setting CPU freq required by perf. This can be useful when measuring performance when CPU frequency scaling is used. I don't know if it's possible now.

Reply:
All are possible, since you are refining arch perf specific code, it's better to leave this issue until you come up a good solution.

Reply:
OK, we can deal with this later